File: ntics/ntics_login.pas
unit ntics_login;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, Buttons, ExtCtrls;

const
  DefaultPrivilege='Guest';
  default_user = 'Anonymous';

type
  TConnectType = (ctLocal,ctTCPIP,ctNetBEUI,ctIPX);

  TIBLogin = class(TForm)
    MainPanel: TPanel;
    btExtended: TSpeedButton;
    chAutoLogin: TCheckBox;
    edUserName: TLabeledEdit;
    edPassword: TLabeledEdit;
    Panel2: TPanel;
    Label4: TLabel;
    SetPatchBase: TSpeedButton;
    Label5: TLabel;
    Label2: TLabel;
    edUserDatabase: TEdit;
    cbConnectType: TComboBox;
    edServerName: TLabeledEdit;
    edRole: TComboBox;
    Panel3: TPanel;
    Button1: TButton;
    Button2: TButton;
    OpenDialog1: TOpenDialog ;
    procedure btExtendedClick(Sender: TObject);
    procedure SetPatchBaseClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    FCharSet:string;
    procedure SetExtended(const Value: boolean);
    function GetUserNameA: string;
    procedure SetUserNameA(const Value: string);
    function GetRole: string;
    function GetDataBase: string;
    procedure SetDataBase(const Value: string);
    function GetServerName: string;
    procedure SerServerName(const Value: string);
    function GetPassword: string;
    procedure SetPassword(const Value: string);
    function GetisAutoLogOn: boolean;
    procedure SetisAutoLogOn(const Value: boolean);
    function GetConnectType: TConnectType;
    procedure SetConnectType(const Value: TConnectType);
  public
    property UserName:string read GetUserNameA write SetUserNameA;
    property Role:string read GetRole;
    property DataBase:string read GetDataBase write SetDataBase;
    property ServerName:string read GetServerName write SerServerName;
    property CharSet:string read FCharSet write FCharSet;
    property Password:string read GetPassword write SetPassword;
    property isAutoLogOn:boolean read GetisAutoLogOn write SetisAutoLogOn;
    property ConnectType:TConnectType  read GetConnectType write SetConnectType;
    function DirectLoginExecute:boolean;
    function LoginExecute:boolean;
    procedure SaveConfig;


  end;

var
  IBLogin:TIBLogin;

implementation
{$R *.dfm}
uses ntics_registration, IniFiles ,ntics_storage;
{ TIBLogin }

procedure TIBLogin.SetExtended(const Value: boolean);
begin
  if Value<>btExtended.Down then
  begin
    btExtended.Down:=Value;
    if Value then
    begin
      Height:=Height + Panel2.Height;
      Panel2.Visible:=true;
    end
    else
    begin
      Panel2.Visible:=false;
      Height:=Height - Panel2.Height;
    end;
  end;
end;


procedure TIBLogin.btExtendedClick(Sender: TObject);
var
  x:boolean;
begin
  x := btExtended.Down;
  btExtended.Down := not x;
  SetExtended(x);
end;

function TIBLogin.GetUserNameA: string;
begin
  Result:=edUserName.Text;
end;

procedure TIBLogin.SetUserNameA(const Value: string);
begin
  edUserName.Text:=Value;
end;

function TIBLogin.GetRole: string;
begin
  Result:=edRole.Text;
end;

procedure TIBLogin.SetPatchBaseClick(Sender: TObject);
begin
  if OpenDialog1.Execute then DataBase := OpenDialog1.FileName;
end;

function TIBLogin.GetDataBase: string;
begin
  Result:=edUserDatabase.Text;
end;

procedure TIBLogin.SetDataBase(const Value: string);
begin
  edUserDatabase.Text:=Value;
end;

function TIBLogin.GetServerName: string;
begin
  Result:=edServerName.Text;
end;

procedure TIBLogin.SerServerName(const Value: string);
begin
  edServerName.Text:=Value;
end;

function TIBLogin.GetPassword: string;
begin
  Result:=edPassword.Text;
end;

procedure TIBLogin.SetPassword(const Value: string);
begin
  edPassword.Text:=Value;
end;

function TIBLogin.GetisAutoLogOn: boolean;
begin
  Result:=chAutoLogin.Checked;
end;

procedure TIBLogin.SetisAutoLogOn(const Value: boolean);
begin
  chAutoLogin.Checked:=Value;
end;

function TIBLogin.GetConnectType: TConnectType;
begin
  Result:= TConnectType(cbConnectType.ItemIndex);
end;

procedure TIBLogin.SetConnectType(const Value: TConnectType);
begin
  cbConnectType.ItemIndex := Integer(Value);
    if Value = ctLocal then
        edServerName.Visible := False
    else
        edServerName.Visible := True;
end;

function TIBLogin.DirectLoginExecute: boolean;
begin
  Result:= (ShowModal = mrOk);
  if Result then SaveConfig;
end;

function TIBLogin.LoginExecute: boolean;
begin
  if not isAutoLogOn then
  begin
    Result:= (ShowModal = mrOk);
    if Result then SaveConfig;
  end
  else Result:=true;
end;

procedure TIBLogin.SaveConfig;
begin
  // пользователь
  LocalStorage.WriteString('IBLogin','CurrentUser',UserName);
  LocalStorage.WriteBool('USER'+UserName,'AutoLogOn',isAutoLogOn);

  // Сохранение начальных параметров пользователя
  LocalStorage.SaveCrypt('USER'+UserName,'CurrentRole',edRole.Text);
  if isAutoLogon then
    LocalStorage.SaveCrypt('USER'+UserName,'Password',edPassword.Text);

  // Сохранение параметров базы
  LocalStorage.WriteString('IBLogin','CharSet',CharSet);
  LocalStorage.WriteInteger('IBLogin','ConnectType',integer(ConnectType));
  LocalStorage.WriteString('IBLogin','SERVERNAME',edServerName.Text);
  LocalStorage.WriteString('IBLogin','UserDataBase',edUserDatabase.Text);
end;

procedure TIBLogin.FormCreate(Sender: TObject);
begin
  // Установка начальных параметров пользователя
  UserName := LocalStorage.ReadString('IBLogin','CurrentUser',default_user);
  isAutoLogOn := LocalStorage.ReadBool('USER'+UserName,'AutoLogOn',false);
  //Role := nt:LoadValue("USER"+UserName,"CurrentRole",default_role);
  if isAutoLogOn then
      Password := localStorage.ReadString('USER'+UserName,'Password','')
  else
      Password := '';
  // Установка начальных параметров базы
  ConnectType := TConnectType(LocalStorage.ReadInteger('IBLogin','ConnectType',0));
  ServerName := LocalStorage.ReadString('IBLogin','SERVERNAME','');
  DataBase := LocalStorage.ReadString('IBLogin','UserDataBase','');
  // Установка кодовой страницы
  CharSet := 'WIN1251';
  SetExtended(false);

 {if (ServerName='') and (DataBasePath='') then
  begin
    if FileExists('ServerConfig.ini') then
    begin
      AssignFile(F,'ServerConfig.ini');
      Reset(F);
      Readln(F,S);
      ServerName:=s;
      Readln(F,S);
      DataBasePath:=s;
    end;
  end;}
end;

end.

File: NTICSLib/source/ntics_login.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "ntics_includes.h"
#pragma hdrstop

#include "ntics_login.h"
#include "ntics_registration.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TIBLogin *IBLogin;
String default_user = "Anonymous";
String default_base = "Uknown";
String default_role = "None";

//---------- START LOAD AND SAVE VALUES -----------------------------
__fastcall TIBLogin::TIBLogin(TComponent* Owner)
        : TForm(Owner)
{
    // Основная база приложения
    ProjectBase = 0;
    // Установка начальных параметров пользователя
    UserName = Registration::LoadValue("IBLogin","CurrentUser",default_user);
    isAutoLogin = StrToInt(Registration::LoadValue("USER"+UserName,"AutoLogOn","0"));
    Role = Registration::LoadValue("USER"+UserName,"CurrentRole",default_role);
    if (isAutoLogin)
      Password = Registration::LoadValue("USER"+UserName,"Password","");
    else
      Password = "";
    // Установка начальных параметров базы
    ConnectType = TConnectType(StrToInt(Registration::LoadValue("IBLogin","ConnectType","0")));
    edServerName->Text = Registration::LoadValue("IBLogin","SERVERNAME","");
    edUserDatabase->Text = Registration::LoadValue("IBLogin","UserDataBase","");
    // Установка кодовой страницы
    CharSet = "WIN1251";
    SetExtended(false);
}

void __fastcall  TIBLogin::SaveConfig(void)
{
    // пользователь
    Registration::SaveValue("IBLogin","CurrentUser",UserName);
    Registration::SaveValue("USER"+UserName,"AutoLogOn",IntToStr(isAutoLogin));
    // Сохранение начальных параметров пользователя
    Registration::SaveValue("USER"+UserName,"CurrentRole",edRole->Text);
    if (isAutoLogin)
        Registration::SaveValue("USER"+UserName,"Password",edPassword->Text,true);

    // Сохранение параметров базы
    //Registration::SaveValue("BASE"+DataBase,"CodeTable",StrToInt(0));
    Registration::SaveValue("IBLogin","ConnectType",IntToStr(ConnectType));
    Registration::SaveValue("IBLogin","SERVERNAME",edServerName->Text);
    Registration::SaveValue("IBLogin","UserDataBase",edUserDatabase->Text);
}

// ---------------------  USER NAME --------------------------------
void __fastcall TIBLogin::SetUserName(String s)
{
    edUserName->Text = s;
}

//----- Дополнительные опции для входа -----
void __fastcall  TIBLogin::SetExtended(bool x)
{
  if (x != btExtended->Down)
  {
    btExtended->Down = x;
    if (x)
    {
      Height = Height + Panel2->Height;
      Panel2->Visible = true;
      edUserName->Enabled = false;
      edPassword->Enabled = false;
    }
    else
    {
      Panel2->Visible = false;
      Height = Height - Panel2->Height;
      edUserName->Enabled = true;
      edPassword->Enabled = true;
    }
  }
}

// Выводим приглашение ввода пароля даже в том случае
// если выбран режим автоматического входа
bool __fastcall TIBLogin::DirectLoginExecute(void)
{
  if (ShowModal() == mrOk)
  {
    SaveConfig();
    return true;
  }
  else return false;
}

// Вход
bool __fastcall TIBLogin::LoginExecute(void)
{
  if (!isAutoLogin)
    if (ShowModal() == mrOk)
    {
      SaveConfig();
      return true;
    }
    else return false;
  return true;
}

//  активация

void __fastcall TIBLogin::SetPatchBaseClick(TObject *Sender)
{
  if (OpenDialog1->Execute()) DataBase = OpenDialog1->FileName;
}
//---------------------------------------------------------------------------


void __fastcall TIBLogin::InitializeBase(TpFIBDatabase *base)
{
    base->Close();

    switch (ConnectType)
    {
      case ctTCPIP : base->DatabaseName = ServerName + ":" + DataBase; break;
      case ctNetBEUI : base->DatabaseName = "\\\\"+ServerName + "\\" + DataBase; break;
      case ctIPX : base->DatabaseName = ServerName + "@" + DataBase; break;
      case ctLocal : base->DatabaseName = DataBase; break;
    }

    base->ConnectParams->UserName = UserName;
    base->ConnectParams->Password = Password;
    base->ConnectParams->CharSet = CharSet;
    base->ConnectParams->RoleName = Role;

    try
    {
        base->Open();
    }
    catch (...)
    {
        throw("Невозможно соединится с базой. Проверьте пароль.");
    }
    ProjectBase = base; // Установка ссылки на основную базу
}

//-------------------  CONNECT TYPE ----------------------------------
TConnectType __fastcall TIBLogin::GetConnectType(void)
{
    return TConnectType(cbConnectType->ItemIndex);
}

void __fastcall TIBLogin::SetConnectType(TConnectType value)
{
    cbConnectType->ItemIndex = value;
    if (value == ctLocal)
        edServerName->Visible = False;
    else
        edServerName->Visible = True;
}



void __fastcall TIBLogin::SetRole(String s)
{
    if (edRole->Items->IndexOf(s)== -1) edRole->Items->Add(s);
    edRole->ItemIndex = edRole->Items->IndexOf(s);
}

//---------------------------------------------------------------------------
void __fastcall TIBLogin::SetAutoLogin(bool x)
{
    chAutoLogin->Checked = x;
}


// переключение в расширенный режим
void __fastcall TIBLogin::btExtendedClick(TObject *Sender)
{
  bool x = btExtended->Down;
  btExtended->Down = !x;
  SetExtended(x);
}


//---------------------------------------------------------------------------
void __fastcall TIBLogin::edUserNameClick(TObject *Sender)
{
    UserName = edUserName->Text;
}


//---------------------------------------------------------------------------
void __fastcall TIBLogin::cbConnectTypeClick(TObject *Sender)
{
   ConnectType = TConnectType(cbConnectType->ItemIndex);
}

